truffle/src/com.oracle.truffle.api.strings/src/com/oracle/truffle/api/strings/AbstractTruffleString.java

```diff
@@ -105,8 +105,17 @@ public abstract class AbstractTruffleString {
      * hashCode value of zero always means that the hash is not calculated yet.
      */
     int hashCode = 0;
+    /**
+     * This array stores the taint information of the {@link TruffleString}s. Taint information is
+     * stored on a per codepoint basis, i.e. for every {@link char} is exactly 1 taint {@link Object}.
+     * <p>
+     * Will be {@code null} for untainted {@link AbstractTruffleString}s.
+     */
+    private Object[] taint;
```

Comment:
I implemented this approach, but instead of using a `byte[]` in the background, I use another `data` field like `AbstractTruffleString#data` (I'm just not allowing another instance of `TaintedString` to be stored in `data`).